packages/appium/lib/appium.js

```diff
@@ -516,6 +524,17 @@ class AppiumDriver extends BaseDriver {
     // handling the command and which didn't
     plugins.length && this.logPluginHandlerReport({cmd, cmdHandledBy});
 
+    // And finally, if the command was createSession, we want to migrate any plugins which were
+    // previously sessionless to use the new sessionId, so that plugins can share state between
+    // their createSession method and other instance methods
```

Comment:
I think I could use some more explanation (though not necessarily in the comment).

- What's a sessionless plugin?
- Does this "migration" of plugins happen once, or multiple times?  
- Why wouldn't this happen in the `createSession()` implementation?

Reply:
> What's a sessionless plugin?

a plugin which exists outside the context of a session. for example, a plugin which is handling the `createSession` command is by definition not associated with any particular session (but then, once the session is created, we want to associate the plugin with that particular session, and not any other sessions that might be happening on the server at once.) a sessionless plugin might also handle `getStatus` or `getSessions`, which are likewise server-level commands and not session-level commands.

> Does this "migration" of plugins happen once, or multiple times?

It should happen on every new session. Which makes me realize there's a bug here, in that sessionless plugins are not "backfilled". I'll need to address that.

> Why wouldn't this happen in the createSession() implementation?

There's no guarantee that Appium's `createSession` will be called, since plugins can override it.

Reply:
> Which makes me realize there's a bug here, in that sessionless plugins are not "backfilled". I'll need to address that.

No, I was wrong, sessionless plugins are recreated as necessary based on the fact that we empty out the list after we 'migrate' them.